addons/binding/org.openhab.binding.rotelra1x/src/main/java/org/openhab/binding/rotelra1x/handler/RotelRa1xHandler.java

```diff
@@ -0,0 +1,350 @@
+            } else if (b == terminator1 || b == terminator2) {
+                break;
+            } else {
+                result.append((char) b);
+            }
+        }
+        return result.toString();
+    }
+
+    public PercentType readVolume() throws IOException {
+        String volumeString = readUntil('!');
+        int volume;
+        if ("min".equals(volumeString)) {
+            volume = 0;
+        } else if ("max".equals(volumeString)) {
+            volume = maximumVolume;
+        } else {
+            volume = Integer.parseInt(volumeString, 10);
+        }
+        double volumePct = volume * 100.0 / maximumVolume;
+        return new PercentType(BigDecimal.valueOf(Math.round(volumePct)));
+    }
+
+    public PercentType readDimmer() throws IOException {
+        String dimmerString = readUntil('!');
+        int dimmer = Integer.parseInt(dimmerString, 10);
+        // Let's make 100 % the brightest, makes more sense
+        double dimmerPct = 100.0 - (dimmer * 100.0 / 6.0);
+        return new PercentType(BigDecimal.valueOf(Math.round(dimmerPct)));
+    }
+
+    public DecimalType readFrequency() throws IOException {
+        String freqString = readUntil('!');
+        double freq;
+        if ("off".equals(freqString)) {
+            freq = 0.0;
+        } else {
+            freq = Double.parseDouble(freqString);
+        }
+        return new DecimalType(freq);
+    }
+
+    void powerOnRefresh() {
+        scheduler.schedule(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    send("get_volume!");
+                    send("get_current_source!");
+                } catch (IOException | ConfigurationError e) {
+                    logger.info("Failed to request volume and source after powering on.", e);
+                }
+            }
+        }, 0, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void run() {
+        while (!exit) {
+            try {
+                if (serialPort == null) {
+                    connect();
+                }
+                String command = readCommand();
+                if ("volume".equals(command)) {
+                    PercentType vol = readVolume();
+                    updateState(getThing().getChannel("volume").getUID(), vol);
+                } else if ("mute".equals(command)) {
+                    String muteState = readUntil('!');
+                    updateState(getThing().getChannel("mute").getUID(),
+                            "on".equals(muteState) ? OnOffType.ON : OnOffType.OFF);
+                } else if ("power_off".equals(command)) {
+                    power = false;
+                    updateState(getThing().getChannel("mute").getUID(), OnOffType.OFF);
+                    updateState(getThing().getChannel("power").getUID(), OnOffType.OFF);
+                    updateState(getThing().getChannel("volume").getUID(), UnDefType.NULL);
```

Comment:
NULL should not be set for channels, this state is only valid at system startup.
I would actually avoid changing volume and source - won't they be still the same after powering on again?

Reply:
They will be the same when powering on, so I've changed it. however if someone sends a command while the unit is off, that won't stick. This is why I set them to NULL, as a "disabled" state. But fine to leave at the current values.